src/rank_llm/retrieve_and_rerank.py

```diff
@@ -22,81 +24,112 @@ def retrieve_and_rerank(
-    if interactive and default_agent is not None:
-        agent = default_agent
-    # Construct Rerank Agent
-    elif "gpt" in model_path or use_azure_openai:
-        openai_keys = get_openai_api_key()
-        agent = SafeOpenai(
-            model=model_path,
-            context_size=context_size,
-            prompt_mode=prompt_mode,
-            num_few_shot_examples=num_few_shot_examples,
-            keys=openai_keys,
-            **(get_azure_openai_args() if use_azure_openai else {}),
-        )
-    elif "vicuna" in model_path.lower() or "zephyr" in model_path.lower():
-        if model_path.lower() == "rank_zephyr":
-            model_full_path = "castorini/rank_zephyr_7b_v1_full"
-        elif model_path.lower() == "rank_vicuna":
-            model_full_path = "castorini/rank_vicuna_7b_v1"
-        else:
-            model_full_path = model_path
+    """Retrieve candidates using Anserini API and rerank them
 
-        print(f"Loading {model_path} ...")
+    Returns:
+        - List of top_k_rerank candidates
+    """
 
-        agent = RankListwiseOSLLM(
-            model=model_full_path,
-            context_size=context_size,
-            prompt_mode=prompt_mode,
-            num_few_shot_examples=num_few_shot_examples,
-            device=device,
-            num_gpus=num_gpus,
-            variable_passages=variable_passages,
-            window_size=window_size,
-            system_message=system_message,
-            vllm_batched=vllm_batched,
+    # Get reranking agent
+    reranker = get_reranker(model_path.lower(), default_agent, interactive, **kwargs)
+
+    # Retrieve initial candidates
+    print(f"Retrieving top {top_k_retrieve} passages...")
+    requests = retrieve(
+        dataset,
+        top_k_retrieve,
+        interactive,
+        retrieval_mode,
+        retrieval_method,
+        query,
+        qid,
+        **kwargs,
+    )
+    print(f"Retrieval complete!")
+
+    # Reranking stage
+    print(f"Reranking and returning {top_k_rerank} passages with {model_path}...")
+    if reranker is None:
+        # No reranker. IdentityReranker leaves retrieve candidate results as is or randomizes the order.
+        shuffle_candidates = True if model_path == "rank_random" else False
+        rerank_results = IdentityReranker().rerank_batch(
+            requests,
+            rank_end=top_k_retrieve,
+            shuffle_candidates=(shuffle_candidates),
         )
     else:
-        raise ValueError(f"Unsupported model: {model_path}")
+        # Reranker is of type RankLLM
+        for pass_ct in range(num_passes):
+            print(f"Pass {pass_ct + 1} of {num_passes}:")
+
+            rerank_results = reranker.rerank_batch(
+                requests,
+                rank_end=top_k_retrieve,
+                rank_start=0,
+                shuffle_candidates=shuffle_candidates,
+                logging=print_prompts_responses,
+                top_k_retrieve=top_k_retrieve,
+                **kwargs,

```

Comment:
Have you tested this cl with unexpected keys in kwargs? I expect passing kwargs here, without changing the `rerank_batch`  would simply flatten the kwargs elements as keyword parameters and raise type error.
Both rerank and rerank_batch functions should take **kwargs as input and properly filter it (similar to what you are doing in the get_reranker and retrive function here).